src/python/gem5/simulate/multi_sim.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2024 The Regents of the University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+import os
+import signal
+from functools import partial
+from time import sleep
+from typing import (
+    Any,
+    Callable,
+    List,
+    Tuple,
+)
+
+from gem5.resources.resource import AbstractResource
+from gem5.utils.multiprocessing import Process
+
+from .simulator import Simulator
+
+
+class MultiSim:
+    """
+    This class is used to run multiple simulations in parallel. It takes a list of
+    callables that return a Simulator object and runs them in parallel. The number of
+    simulations that can be run in parallel is limited to the value of the num_cpus parameter
+    """
+
+    def __init__(self, sim_callables: List[Callable[[], Simulator]]):
+        """
+        :param sim_callables: A list of callables that return a Simulator object
+        """
+        self.sim_callables = sim_callables
+
+    def run_all(self, num_cpus: int):
+        """
+        Run all the simulations in parallel
```

Comment:
Be careful with the phrasing here. You are not always running all the simulations in parallel. If `num_cpus` (`num_proc`/`num_processes`) were set to 1 this statement would not be true, for example.

Reply:
Yeah, you're also not running *all* in parallel. You're running all of the simulations with the parallelism specified with `num_cpus` (should be `num_processes`)